### ISSUE TITLE
Can't connect QueueManager

### ISSUE THREAD

**Issue by ToNorthX** [dsuch/pymqi#78]
@dsuch Hi 
This error occurs after installing MQ7.5:
Can't connect QueueManager .reason [MQI Error. Comp 2,Reson 2380: FAILED: MQRC_SCO_ERROR]
What is the cause of this?

**Comment by JoshMcMQ** [comment]
Can you show the code that causes this error?  What version of MQ did you have installed prior to this?  What version of pymqi are you using?  What version of Python?

**Comment by tongwandouxiong** [comment]
@JoshMcMQ Hi:
The MQ version installed before this is 8.0, the pymqi version is 1.8, and the Python is 2.7. During the installation of MQ, I executed the rpm -ivh *.rpm command and installed all the rpm installation packages.

**Comment by JoshMcMQ** [comment]
Some things you forgot to provide:
1. Can you show the code that causes this error?
2. Was the only component changed the version of MQ.
3. Check and make sure that the location of the key repository is correct.

**Comment by tongwandouxiong** [comment]
![k59 mmu5grx72m 0uisp](https://user-images.githubusercontent.com/34308472/46595431-e496ce00-cb0a-11e8-954d-12beb8096525.png)
![t0703 7z7jvg4wb5qd5ieih](https://user-images.githubusercontent.com/34308472/46595447-f5dfda80-cb0a-11e8-8d00-21476ca673be.png)

**Comment by JoshMcMQ** [comment]
@dsuch this is another instance similar to the MQCSP, if the user does not pass a SCO structure, passing in a default structure is providing no benefit.  The IBM's `MQCNO_DEFAULT` sets `SSLConfigPtr` to `NULL`.  In this case if pymqi didn't pass in the default SCO structure, @tongwandouxion's use case of having pymqi compiled against MQ v8 but then running it with a MQ 7.5 install would likely have worked under IBM's statement I quoted in my prior comment: "though they might work in practice".

Below I looked at both `__init__.py` and `pymqe.c` and I tried to describe a few minor updates that would accomplish this change so that if a `SCO` is not specifically created and passed to `connect_with_options` that it would not pass a default `SCO` structure to the underlying `pymqe.MQCONNX`.  Based on my review I do not believe the `pymqe.MQCONNX` would require any changes to make this work although a suggested a few changes that I think would make it more clear.

---

First, looking at the `connect_with_options` in `__init__.py`:

```
        osco = SCO()
        if 'opts' in kw:
            options = kw['opts']
        if 'cd' in kw:
            ocd = kw['cd']
        if 'sco' in kw:
            rv = pymqe.MQCONNX(name, options, ocd.pack(), user_password, kw['sco'].pack())
        else:
            rv = pymqe.MQCONNX(name, options, ocd.pack(), user_password, osco.pack())
```

I think you can simplify it to this:

```
        #osco = SCO()
        if 'opts' in kw:
            options = kw['opts']
        if 'cd' in kw:
            ocd = kw['cd']
        if 'sco' in kw:
            rv = pymqe.MQCONNX(name, options, ocd.pack(), user_password, kw['sco'].pack())
        else:
            rv = pymqe.MQCONNX(name, options, ocd.pack(), user_password)
```

---

In `pymqe.c` in the `pymqe_MQCONNX` function it does the following, my understanding of this is that the `PyArg_ParseTuple` will leave the `sco` and `sco_len` set to the values they were initialized with.  The subsequent check of `sco` will also be fine if no `sco` was passed and `sco = NULL` since the if statement would not eval to true or even call `checkArgSize` if `sco = NULL`, this would avoid the `return NULL`.

```
#ifdef PYMQI_FEATURE_SSL
  char *sco = NULL;
  int sco_len = 0;
#if PY_MAJOR_VERSION==2
  if (!PyArg_ParseTuple(args, "sls#O|s#", &name, &options, &mqcd, &mqcd_buf_len, &user_password, &sco, &sco_len)) {
#else
  if (!PyArg_ParseTuple(args, "yly#O|y#", &name, &options, &mqcd, &mqcd_buf_len, &user_password, &sco, &sco_len)) {
#endif
    return 0;
  }
  if (sco && checkArgSize(sco_len, PYMQI_MQSCO_SIZEOF, "MQSCO")) {
    return NULL;
  }
#else
```

---

Further down it does the following:

```
#ifdef PYMQI_FEATURE_SSL
  cno.Version = MQCNO_VERSION_5;
  cno.SSLConfigPtr =  (MQSCO *)sco;
#else
  cno.Version = MQCNO_VERSION_2;
#endif
```
If the statement `cno.SSLConfigPtr =  (MQSCO *)sco;` will still resolve to NULL then this is fine since `SSLConfigPtr` would be NULL to begin with.

You could change it to something like this so it is a little more clear what is going on:

```
#ifdef PYMQI_FEATURE_SSL
  if (sco) {
    cno.SSLConfigPtr =  (MQSCO *)sco;
  }
  cno.Version = MQCNO_VERSION_5;
#else
  cno.Version = MQCNO_VERSION_2;
#endif
```

---

Also note that technically speaking the `SSLConfigPtr` was introduced at `MQCNO_VERSION_4`, but the subsequent `MQCSP` setting for `user` and `pass` does require `MQCNO_VERSION_5` for `SecurityParmsPtr` and is relying on it being set to five above.

You could further clarify this with:

```
#ifdef PYMQI_FEATURE_SSL
  if (sco) {
    cno.SSLConfigPtr =  (MQSCO *)sco;
  }
  cno.Version = MQCNO_VERSION_4;
#else
  cno.Version = MQCNO_VERSION_2;
#endif
```

And then in further down change:

```
    else {
      cno.SecurityParmsPtr = &csp;
    }
```

to

```
    else {
      cno.Version = MQCNO_VERSION_5;
      cno.SecurityParmsPtr = &csp;
    }
```
---

I apologize I am a MQ admin and not much of a developer don't quite know the ins and outs of doing a pull request, etc, but wanted to provide my feedback if you want to take it or not is up to you :).

**Comment by tongwandouxiong** [comment]
@
I understand some of your meaning, that is to say, without passing SCO, pymqi needs to be compiled according to the installed version of MQ, and I also tried to reinstall MQv7.5 and successfully connected to the queue manager. At this point, when I deploy the application to linux suse, this problem still occurs. Suse is installed as MQ v7.5, python v 2.7 
Exception information:
![46595431-e496ce00-cb0a-11e8-954d-12beb8096525](https://user-images.githubusercontent.com/34308472/47077340-4fdf4f00-d233-11e8-8c97-55b8f6c508ca.png)
Code:

![46595431-e496ce00-cb0a-11e8-954d-12beb8096525](https://user-images.githubusercontent.com/34308472/47078090-160f4800-d235-11e8-83e1-8025cb49d873.png)

**Comment by JoshMcMQ** [comment]
@tongwandouxiong For the Suse failure did you compile the version you are using against MQ v7.5, if so it does not make sense that it would fail.

**Comment by SeyfSV** [comment]
Fixed in #170
